backend/listing/views.py

```diff
@@ -0,0 +1,101 @@
+import json
+
+from django_filters.rest_framework import DjangoFilterBackend
+from drf_spectacular.utils import extend_schema
+from rest_framework import permissions
+from rest_framework import viewsets, status
+from rest_framework.response import Response
+
+from .filters import ListingFilter
+from .models import Listing
+from .serializers import ListingSerializer
+
+
+@extend_schema(tags=['Listing'])
+class ListingViewSet(viewsets.ModelViewSet):
+    """
+    Viewset for API endpoint that implements CRUD operations for listing(cards for sale).
+    """
+
+    queryset = Listing.objects.all().order_by('id')
+    serializer_class = ListingSerializer
+    permission_classes = [permissions.IsAuthenticated]
+    filter_backends = [DjangoFilterBackend]
+    filterset_class = ListingFilter
+
+    def get_queryset(self):
+        return self.queryset.filter(user=self.request.user)
+
+    def perform_create(self, serializer):
+        serializer.save(user=self.request.user)
+
+    def update(self, request, *args, **kwargs):
```

Comment:
The `update` endpoint should be dedicated to just generic fields updating.

The `buy` functionality deserves its own endpoint because it is a different type of logic. In the future, the buy functionality might trigger other processes, like creating an order, sending emails, etc. If this is in the `update` endpoint, you can imagine how bad it will get eventually, so that's why it would be ideal if we can put it in a separate endpoint.

Do:
`/api/listing/<id>/buy`

Don't:
`/api/listing/buy?id=2`

Reply:
Notes taken,
I have re-implemented the logic following the suggestions

I have seperated the functionallity to 3 endpoints:
**PUT** request to - `/api/listing/<id>/buy/` - will update the listing as sold and make it unlisted
**GET** request to - `/api/listing/sold/` - will return all *sold* listings
**GET** request to - `/api/listing/unsold/` - will return all *unsold* listings

>[!Important]
>I have made а custom permission class to check if the user is the same as the owner which is trying to mark the listing as sold. This can be used for any of the other ViewSets as well.

Please let me know if I need to change the quering for sold and unsold listings as a query parameter to the `/api/listing/?is_sold=true` endpoint. I was not sure which way to choose. 

Also I used the PUT method for this endpoint `/api/listing/<id>/buy/` because I think it is more intuitive since it updates an existing object rather than creating new object which in my opinion is a job for the POST method 🙂

Reply:
I would generally go for a query parameter, since you only filter the data, and you don't do any more actions. However, it's okay for now, no need to fix it at this point.

Reply:
Okay, thank you for the feedback, I will wait for the listing task to be ready and I will correct this accordingly.